js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala

```diff
@@ -280,30 +289,52 @@ abstract class AbstractNodeJSEnv(
     }
 
     def close(): Unit = {
+      /* Close the socket first. This will cause any existing and upcoming
+       * calls to `awaitConnection()` to be canceled and throw a
+       * `SocketException` (unless it has already successfully completed the
+       * `accept()` call).
+       */
       serverSocket.close()
-      if (jvm2js != null)
-        jvm2js.close()
-      if (js2jvm != null)
-        js2jvm.close()
-      if (comSocket != null)
-        comSocket.close()
+
+      /* Now wait for a possibly still-successful `awaitConnection()` to
+       * complete before closing the sockets.
+       */
+      synchronized {
+        if (jvm2js != null)
+          jvm2js.close()
+        if (js2jvm != null)
+          js2jvm.close()
+        if (comSocket != null)
+          comSocket.close()
+      }
     }
 
     /** Waits until the JS VM has established a connection or terminates
      *
      *  @return true if the connection was established
      */
-    private def awaitConnection(): Boolean = {
+    private def awaitConnection(): Boolean = synchronized {
       serverSocket.setSoTimeout(acceptTimeout)
+
       while (comSocket == null && isRunning) {
-        try {
-          comSocket = serverSocket.accept()
-          jvm2js = new DataOutputStream(
-            new BufferedOutputStream(comSocket.getOutputStream()))
-          js2jvm = new DataInputStream(
-            new BufferedInputStream(comSocket.getInputStream()))
+        val comSocket0 = try {
+          serverSocket.accept()
         } catch {
-          case to: SocketTimeoutException =>
+          case to: SocketTimeoutException => null
+        }
+
+        if (comSocket0 != null) {
+          val jvm2js0 = new DataOutputStream(
+              new BufferedOutputStream(comSocket0.getOutputStream()))
+          val js2jvm0 = new DataInputStream(
+              new BufferedInputStream(comSocket0.getInputStream()))
+
+          /* Assign those three fields together, without the possibility of an
+           * exception happening in the middle (see #3408).
+           */
+          comSocket = comSocket0
+          jvm2js = jvm2js0
+          js2jvm = js2jvm0
         }
```

Comment:
Shouldn't this go outside of the while loop? Have you considered putting the accept call in a tail recursive helper instead?

Reply:
I more meant something like this:

```scala
@tailrec
def waitAccept(): Option[Socket] = {
  if (!isRunning) {
    None
  } else {
    try {
      Some(serverSocket.accept())
    } catch {
      case _: SocketTimeoutException => waitAccept()
    }
  }
}

val optCom = waitAccept()

optCom.foreach {
  // Create streams and assign.
}

optCom.isDefined
```

IMO the exact conditions when what happens are clearer that way. Also the assignment is not in the loop anymore; it only can happen in the last iteration (if at all).

Reply:
OK, I see. Better that way, indeed. Done.